usr/man/_mips.d/man3/ftw.3
.\" SCCSID: @(#)ftw.3	8.2	3/28/91
.TH ftw 3
.SH Name
ftw \- walk a file tree
.SH Syntax
.B #include <ftw.h>
.PP
.B int ftw (\fIpath, fn, depth\fP)
.br
.B char *\fIpath\fP;
.br
.B int (*\fIfn) ( )\fP;
.br
.B int \fIdepth\fP;
.SH Description
.NXR "ftw subroutine"
.NXR "directory" "descending tree"
The
.PN ftw
subroutine recursively descends the directory hierarchy
rooted in
.IR path .
For each object in the hierarchy,
.PN ftw
calls
.IR fn ,
passing it a pointer to a null-terminated
character string containing the name of the object,
a pointer to a
.B stat
structure
containing information about the object, and an integer.
For further information, see 
.MS stat 2 .
Possible values of the integer, defined in the <ftw.h> header file,
are FTW_F for a file, FTW_D for a directory, FTW_DNR for
a directory that cannot be read, and FTW_NS for an object
for which
.B stat
could not successfully be executed.
If the integer is FTW_DNR,
descendants of that directory will not be processed.
If the integer is FTW_NS, the the contents of the
.B stat
structure will be undefined. 
An example of an object that would cause
FTW_NS to be passed to
.I fn\^
would be a file in a directory
with read but without execute (search) permission.
.PP
The
.PN ftw
subroutine visits a directory before visiting any of its descendants.
.PP
The tree traversal continues until the tree is exhausted,
an invocation of
.I fn\^
returns a nonzero value,
or some error is detected within
.PN ftw
(such as an I/O error).
If the tree is exhausted,
.PN ftw
returns zero.
If
.I fn\^
returns a nonzero value,
.PN ftw
stops its tree traversal and returns whatever
value was returned by
.IR fn .
If
.PN ftw
detects an error, it returns
\-1, and sets the error type in
.IR errno .
.PP
The
.PN ftw
subroutine
uses one file descriptor for each level in the tree.
The
.I depth\^
argument limits the number of file descriptors so used.
If
.I depth\^
is zero or negative, the effect is the same as if it were 1.
The
.I depth\^
must not be greater than the number of file descriptors currently
available for use.
The
.PN ftw
subroutine will run more quickly if
.I depth\^
is at least as large as the number of levels in the tree.
.SH Restrictions
Because
.PN ftw
is recursive, it is possible for it to terminate with a memory
fault when applied to very deep file structures.
.br
It could be made to run faster and use less storage on deep
structures at the cost of considerable complexity.
.br
The
.PN ftw
subroutine uses 
.MS malloc 3
to allocate dynamic storage during its operation.
If
.PN ftw
is forcibly terminated, such as by
.I longjmp\^
being executed by
.I fn\^
or an interrupt routine,
.PN ftw
will not have a chance to free that storage,
so it will remain permanently allocated.
A safe way to handle interrupts is to store
the fact that an interrupt has occurred,
and arrange to have
.I fn\^
return a nonzero value at its next invocation.
.SH Diagnostics
.TP 15
[EACCES] 
Search permission is denied on a component of \fIpath\fP or read
permission is denied for path.
.TP
[ENAMETOOLONG]
The length of the path string exceeds {PATH_MAX}, or
a pathname component is longer than {NAME_MAX}.
.TP
[ENOENT]
The path argument points to the name of a file which
does not exist, or to an empty string and the environment
defined is POSIX or SYSTEM_FIVE.
.TP
[ENOTDIR]
A component of \fIpath\fP is not a directory.
.TP
[ENOMEM]
Not enough memory was available to complete the file tree
walk.
.SH See Also
stat(2), malloc(3)


lib/libc/gen/ftw.c
#ifndef lint
static	char	*sccsid = "@(#)ftw.c	4.1	(ULTRIX)	7/3/90";
#endif lint

/************************************************************************
 *									*
 *			Copyright (c) 1985 by				*
 *		Digital Equipment Corporation, Maynard, MA		*
 *			All rights reserved.				*
 *									*
 *   This software is furnished under a license and may be used and	*
 *   copied  only  in accordance with the terms of such license and	*
 *   with the  inclusion  of  the  above  copyright  notice.   This	*
 *   software  or  any  other copies thereof may not be provided or	*
 *   otherwise made available to any other person.  No title to and	*
 *   ownership of the software is hereby transferred.			*
 *									*
 *   This software is  derived  from  software  received  from  the	*
 *   University    of   California,   Berkeley,   and   from   Bell	*
 *   Laboratories.  Use, duplication, or disclosure is  subject  to	*
 *   restrictions  under  license  agreements  with  University  of	*
 *   California and with AT&T.						*
 *									*
 *   The information in this software is subject to change  without	*
 *   notice  and should not be construed as a commitment by Digital	*
 *   Equipment Corporation.						*
 *									*
 *   Digital assumes no responsibility for the use  or  reliability	*
 *   of its software on equipment which is not supplied by Digital.	*
 *									*
 ************************************************************************/

/************************************************************************
 *			Modification History
 *
 * 001	David L Ballenger, 22-Oct-1985
 *	Copy filename so that destination string is null terminated.
 *
 ************************************************************************/

/*LINTLIBRARY*/
/***************************************************************
 *	ftw - file tree walk
 *
 *	int ftw (path, fn, depth)  char *path; int (*fn)(); int depth;
 *
 *	Given a path name, ftw starts from the file given by that path
 *	name and visits each file and directory in the tree beneath
 *	that file.  If a single file has multiple links within the
 *	structure, it will be visited once for each such link.
 *	For each object visited, fn is called with three arguments.
 *	The first contains the path name of the object, the second
 *	contains a pointer to a stat buffer which will usually hold
 *	appropriate information for the object and the third will
 *	contain an integer value giving additional information about
 *
 *		FTW_F	The object is a file for which stat was
 *			successful.  It does not guarantee that the
 *			file can actually be read.
 *
 *		FTW_D	The object is a directory for which stat and
 *			open for read were both successful.
 *
 *		FTW_DNR	The object is a directory for which stat
 *			succeeded, but which cannot be read.  Because
 *			the directory cannot be read, fn will not be
 *			called for any descendants of this directory.
 *
 *		FTW_NS	Stat failed on the object because of lack of
 *			appropriate permission.  This indication will
 *			be given, for example, for each file in a
 *			directory with read but no execute permission.
 *			Because stat failed, it is not possible to
 *			determine whether this object is a file or a
 *			directory.  The stat buffer passed to fn will
 *			contain garbage.  Stat failure for any reason
 *			other than lack of permission will be
 *			considered an error and will cause ftw to stop 
 *			and return -1 to its caller.
 *
 *	If fn returns nonzero, ftw stops and returns the same value
 *	to its caller.  If ftw gets into other trouble along the way,
 *	it returns -1 and leaves an indication of the cause in errno.
 *
 *	The third argument to ftw does not limit the depth to which
 *	ftw will go.  Rather, it limits the depth to which ftw will
 *	go before it starts recycling file descriptors.  In general,
 *	it is necessary to use a file descriptor for each level of the
 *	tree, but they can be recycled for deep trees by saving the
 *	position, closing, re-opening, and seeking.  It is possible
 *	to start recycling file descriptors by sensing when we have
 *	run out, but in general this will not be terribly useful if
 *	fn expects to be able to open files.  We could also figure out
 *	how many file descriptors are available and guarantee a certain
 *	number to fn, but we would not know how many to guarantee,
 *	and we do not want to impose the extra overhead on a caller who
 *	knows how many are available without having to figure it out.
 *
 *	It is possible for ftw to die with a memory fault in the event
 *	of a file system so deeply nested that the stack overflows.
 **************************************************************/

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/dir.h>
#include <errno.h>
#include <ftw.h>

#define NULL 0

extern char *malloc(), *strcpy();
extern long lseek();
extern int errno;

int
ftw(path, fn, depth)
char *path;
int (*fn)();
int depth;
{
	int rc, n;
	DIR *fd;
	char *subpath, *component;
	struct stat sb;
	struct direct *dirp;

	/* Try to get file status.
				If unsuccessful, errno will say why. */
	if(stat(path, &sb) < 0)
		return(errno == EACCES? (*fn)(path, &sb, FTW_NS): -1);

	/*
	 *	The stat succeeded, so we know the object exists.
	 *	If not a directory, call the user function and return.
	 */
	if((sb.st_mode & S_IFMT) != S_IFDIR)
		return((*fn)(path, &sb, FTW_F));

	/*
	 *	The object was a directory.
	 *
	 *	Open a file to read the directory
	 */
	fd = opendir(path);

	/*
	 *	Call the user function, telling it whether
	 *	the directory can be read.  If it can't be read
	 *	call the user function or indicate an error,
	 *	depending on the reason it couldn't be read.
	 */
	if(fd == NULL)
		return(errno == EACCES? (*fn)(path, &sb, FTW_DNR): -1);

	/* We could read the directory.  Call user function. */
	rc = (*fn)(path, &sb, FTW_D);
	if(rc != 0)
		return(rc);

	/* Allocate a buffer to hold generated pathnames. */
	n = strlen(path);
	subpath = malloc((unsigned)(n+MAXNAMLEN+2));
	if(subpath == NULL) {
		closedir(fd);
		errno = ENOMEM;
		return(-1);
	}
	
	/* Create a prefix to which we will append component names */
	(void)strcpy(subpath, path);
	if(subpath[0] != '\0' && subpath[n-1] != '/')
		subpath[n++] = '/';
	component = &subpath[n];

	/*
	 *	Read the directory one component at a time.
	 *	We must ignore "." and "..", but other than that,
	 *	just create a path name and call self to check it out.
	 */
	while((dirp = readdir(fd)) != NULL) {
		if( strcmp(dirp->d_name, ".")  != 0 &&
		    strcmp(dirp->d_name, "..") != 0
		  ) {
			int i;
			char *p, *q;
			long here;

			/* Append component name to the working path */
			p = component;
			q = dirp->d_name;
			while (*p++ = *q++);

			/*
			 *	If we are about to exceed our depth,
			 *	remember where we are and close a file.
			 */
			if(depth <= 1) {
				here = telldir(fd);
				closedir(fd);
			}

			/*
			 *	Do a recursive call to process the file.
			 *	(watch this, sports fans)
			 */
			rc = ftw(subpath, fn, depth-1);
			if(rc != 0) {
				free(subpath);
				if(depth > 1)
					closedir(fd);
				return(rc);
			}

			/*
			 *	If we closed the file, try to reopen it.
			 */
			if(depth <= 1) {
				fd = opendir(path);
				if(fd == NULL) {
					free(subpath);
					return(-1);
				}
				seekdir(fd, here);
			}
		}
	}

	/*
	 *	We got out of the subdirectory loop.  Clean up.
	 */
	free(subpath);
	closedir(fd);
	return(0);
}
